README.md
# to_precision

[![Travis branch](https://img.shields.io/travis/epmoyer/to-precision/overbar.svg)](https://travis-ci.org/epmoyer/to-precision)

Formatting floating point numbers to standard, scientific, or engineering notation with a specified number of significant digits.

Python: 2.7, 3.3, 3.4, 3.5, 3.6

Created by:  
**William Rusnack**
[GitHub](github.com/BebeSparkelSparkel)
[LinkedIn](linkedin.com/in/williamrusnack/)
[Email](williamrusnack@gmail.com)  
**Eric Moyer**
[GitHub](github.com/epmoyer)
[Email](eric@lemoncrab.com)  
**Randle Taylor**
[GitHub](github.com/randlet)


## Install

    pip install git+https://github.com/BebeSparkelSparkel/to-precision.git

### standard notation
to_precision.**std_notation**(value, precision)

standard notation (US version). http://www.mathsisfun.com/definitions/standard-notation.html

returns a string of value with the proper precision

ex:

    >>> std_notation(5, 2)
    '5.0'
    >>> std_notation(5.36, 2)
    '5.4'
    >>> std_notation(5360, 2)
    '5400'
    >>> std_notation(0.05363, 3)
    '0.0536'

### scientific notation
to_precision.**sci_notation**(value, precision, delimiter='e'):

scientific notation. https://www.mathsisfun.com/numbers/scientific-notation.html

returns a string of value with the proper precision and 10s exponent
delimiter is placed between the decimal value and 10s exponent

ex:

    >>> sci_notation(123, 1, 'E')
    '1E2'
    >>> sci_notation(123, 3, 'E')
    '1.23E2'
    >>> sci_notation(.126, 2, 'E')
    '1.3E-1'

### engineering notation
to_precision.**eng_notation**(value, precision, delimiter='e')

engineering notation. http://www.mathsisfun.com/definitions/engineering-notation.html

returns a string of value with the proper precision and 10s exponent that is divisible by 3
delimiter is placed between the decimal value and 10s exponent

ex:

    >>> eng_notation(123, 1, 'E')
    '100E0'
    >>> eng_notation(1230, 3, 'E')
    '1.23E3'
    >>> eng_notation(.126, 2, 'E')
    '120E-3'

### auto notation
to_precision.**auto_notation**(value, precision, delimiter='e')

Automatically selects between standard notation (US version) and scientific notation.
Values in the range 0.001 < abs(value) < 1000 return standard notation.

http://www.mathsisfun.com/definitions/standard-notation.html
https://www.mathsisfun.com/numbers/scientific-notation.html

returns a string of value with the proper precision

ex:

    >>> auto_notation(123, 4)
    '123.4'
    >>> std_notation(1234, 4)
    '1.234e3'

### to_precision
to_precision.**to_precision**(value, precision, notation='auto', delimiter='e', auto_limit=3, strip_zeros=False, preserve=False)  

Converts a value to the specified notation and precision
value - any type that can be converted to a float
precision - integer that is greater than zero
notation - string

* value
    * The number to convert
* precision
    * The digits of precision
* notation
    * 'auto' - selects standard notation when abs(power) < auto_limit else returns scientific notation.
    * 'sci' or 'scientific' - returns scientific notation. https://www.mathsisfun.com/numbers/scientific-notation.html
    * 'eng' or 'engineering' - returns engineering notation. http://www.mathsisfun.com/definitions/engineering-notation.html
    * 'std' or 'standard' - returns standard notation. http://www.mathsisfun.com/definitions/standard-notation.html
* delimiter
    * Text that is placed between the decimal value and 10s exponent
* auto_limit
    * Integer. When abs(power) exceeds this limit, 'auto'
    mode will return scientific notation. The default (3) will cause 'auto' mode to return scientific notation for 0.001 < abs(value) < 1000
* strip_zeros
    * If true, trailing decimal zeros will be removed.
* preserve_integer
    * If true, 'std' will preserve all digits when returning
    values that have no decimal component.
* convention
    * See "Significance notation convention" section below
        * 'dot' - selects dot convention.
        * 'overbar' - selects overbar convention.

## Significance notation convention

In a few situations, differentiating which _printed_ digits are significant and which are not can be _ambiguous_. 

In these cases this library defaults to ``'dot'`` notation.

``'overbar'`` notation and ``'none'`` are also supported via the ``to_precision()`` function's ``convention`` argument.

Potentially ambiguous notations arise in two cases

* Standard notation, where the result is an integer that ends in one or more zeros.
* Engineering notation, where the value before the exponent in the result is an integer that ends in one or more zeros.


The following examples illustrate the behavior of the 3 convention options in these situations.

![overbar](https://github.com/epmoyer/to-precision/raw/overbar/img/overbar_example.png)
    
See https://en.wikipedia.org/wiki/Significant_figures#Significant_figures_rules_explained

**NOTE:**

The overbar character is rendered using a Unicode COMBINING OVERLINE (U+0305). in Python 2, calling to_precision() with convention='overbar' will **ALWAYS** return a unicode() object, **WHETHER OR NOT** the result contains an overbar (to avoid the surprise of only _occasionally_ returning Unicode).

## Implicit vs. Explicit precision

Consider the following data set:

    A                 B              C              D              
    ----------------  -------------  -------------  -------------  
    1.00000000000     0.98765432100  0.08700000000  0.00000000234  
    1.20000000000     1.40000000000  1.00000000000  0.00000000002  
    1234.00000000000  1.23450000000  0.02345678000  0.00000000000 

In a scientific or engineering context, one typically indicates precision *explicitly* by showing all significant digits.  This can be accomplished using `to_pecision()` with its default arguments.

Using `to_precision(x, 3)` to render the example data set yields:

    A       B      C       D         
    ------  -----  ------  --------  
    1.00    0.988  0.0870  2.34e-9   
    1.20    1.40   1.00    2.00e-11  
    1.23e3  1.23   0.0235  0.00

In a business or informal context one may wish to *implicitly* specify precision, and render results in their simplest, least cluttered form given that precision.  This can be accomplished by invoking the `strip_zeros` and `preserve_integer` arguments.  Additionally, `auto_limit` can be used to control the threshold at which scientific notation is adopted.

Using `to_precision(x, 3, auto_limit=4, strip_zeros=True, preserve_integer=True)` to render the example data set yields:

    A     B      C       D        
    ----  -----  ------  -------  
    1     0.988  0.087   2.34e-9  
    1.2   1.4    1       2e-11    
    1234  1.23   0.0235  0 

## Decimal Notation

to-precision uses decimal notation to indicate precision in cases where the result:

* Is an integer
* Ends with one or more zeros
* All ending zeros are significant

See: https://en.wikipedia.org/wiki/Significant_figures#Significant_figures_rules_explained

For examlple:

    >>> to_precision(120, 2)
    '120'                      # Zero is not significant
    >>> to_precision(120, 3)
    '120.'                     # Zero is significant
    >>> to_precision(100, 2)
    '100'                      # First zero is significant but second is not
    >>>


## Demonstration

The script `demonstration.py` demonstrates the behavior of the various options...

    $./demonstration.py 
    Default (Auto Notation):
        to_precision(value, precision)

        value         precision=1    precision=2    precision=3    precision=4    precision=5
        ------------  -------------  -------------  -------------  -------------  -------------
        0             0              0.0            0.00           0.000          0.0000
        1             1              1.0            1.00           1.000          1.0000
        10            10             10.            10.0           10.00          10.000                                           
        ...
        (many cases)
        ...
        0.000123456  0.0001         0.00012        0.000123       0.0001235      0.00012346
        0.001234567  0.001          0.0012         0.00123        0.001235       0.0012346
        0.012345678  0.01           0.012          0.0123         0.01235        0.012346
        0.123456789  0.1            0.12           0.123          0.1235         0.12346
                  
                                                                                                                                  

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## Fixed
-  Made `demonstration.py` Python 2 compatible (emoved `end` keyword from print statements)

### Added
- `convention` argument to `to_precision()`
- Travis CI integration
- Header comments & permissions so test.py could be run as a script ($ ./tests.py)
- CHANGLELOG.md


## 2.0.0 - 2015-12-03

### Fixed
- `to_precision(0, 1, 'auto') => '0'` instead of '0.e0'`
- `to_precision(0, 1, 'std') => '0'` instead of `'0.'`

### Added
- Options to to_precision(): `auto_limit`, `strip_zeros`, `preserve_integer`
- `demonstration.py`
- `auto_notation()`
- 'delimiter' as default option to `sci_notation()`, `eng_notation()`

### Changed
- `filler` argument to `delimiter`



tests.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

__author__ = 'William Rusnack github.com/BebeSparkelSparkel linkedin.com/in/williamrusnack williamrusnack@gmail.com'

import sys
import unittest
from io import StringIO
from contextlib import contextmanager
from to_precision import to_precision, auto_notation, std_notation, sci_notation, eng_notation, _place_dot, _number_profile
import demonstration

class TestToPrecision(unittest.TestCase):
  def test_auto(self):
    # Automatic conversion of input type to float
    self.assertEqual(
        to_precision("1.2", 2),
        '1.2'
      )

    # within limit
    self.assertEqual(
        to_precision(-552, 2, notation='auto', delimiter='e'),
        '-550'
      )

    self.assertEqual(
        to_precision(552, 2, notation='auto', delimiter='e'),
        '550'
      )

    # over limit
    self.assertEqual(
        to_precision(1001, 2, notation='auto', delimiter='e'),
        '1.0e3'
      )

    self.assertEqual(
        to_precision(-1001, 2, notation='auto', delimiter='e'),
        '-1.0e3'
      )

    self.assertEqual(
        to_precision(0.0009, 2, notation='auto', delimiter='e'),
        '9.0e-4'
      )

    self.assertEqual(
        to_precision(-0.0009, 2, notation='auto', delimiter='e'),
        '-9.0e-4'
      )

    # Auto limit
    self.assertEqual(
        to_precision(1234, 4, notation='auto', delimiter='e'),
        '1.234e3'
      )
    self.assertEqual(
        to_precision(1234, 4, notation='auto', delimiter='e', auto_limit=4),
        '1234'
      )

    self.assertEqual(
        to_precision(.001, 3, notation='auto', delimiter='e'),
        '1.00e-3'
      )
    self.assertEqual(
        to_precision(.001, 3, notation='auto', delimiter='e', auto_limit=4),
        '0.00100'
      )

    self.assertEqual(
        to_precision(12345, 5, notation='auto', delimiter='e', auto_limit=4),
        '1.2345e4'
      )
    self.assertEqual(
        to_precision(12345, 5, notation='auto', delimiter='e', auto_limit=5),
        '12345'
      )

    self.assertEqual(
        to_precision(.000123, 3, notation='auto', delimiter='e', auto_limit=4),
        '1.23e-4'
      )
    self.assertEqual(
        to_precision(.000123, 3, notation='auto', delimiter='e', auto_limit=5),
        '0.000123'
      )

   # preserve_integer
    self.assertEqual(
      to_precision(12345, 2, notation='auto', preserve_integer=False, auto_limit=5),
      '12000')
    self.assertEqual(
      to_precision(12345, 2, notation='auto', preserve_integer=True, auto_limit=5),
      '12345')

    # Zero stripping
    self.assertEqual(
      to_precision(12000000, 5, notation='auto', strip_zeros=False),
      '1.2000e7')
    self.assertEqual(
      to_precision(12000000, 5, notation='auto', strip_zeros=True),
      '1.2e7')
    self.assertEqual(
      to_precision(12000000, 5, notation='sci', strip_zeros=False),
      '1.2000e7')
    self.assertEqual(
      to_precision(12000000, 5, notation='sci', strip_zeros=True),
      '1.2e7')
    self.assertEqual(
      to_precision(12000000, 5, notation='eng', strip_zeros=False),
      '12.000e6')
    self.assertEqual(
      to_precision(12000000, 5, notation='eng', strip_zeros=True),
      '12e6')
    self.assertEqual(
      to_precision(12, 5, notation='std', strip_zeros=False),
      '12.000')
    self.assertEqual(
      to_precision(12, 5, notation='std', strip_zeros=True),
      '12')

  def test_multi(self):
    self.assertEqual(
        to_precision(500, 2, notation='std', delimiter='e'),
        '500'
      )

    self.assertEqual(
        to_precision(500, 2, notation='sci', delimiter='e'),
        '5.0e2'
      )

    self.assertEqual(
        to_precision(1100, 2, notation='eng', delimiter='e'),
        '1.1e3'
      )

class TestAutoNotation(unittest.TestCase):
  def test_multi(self):
    self.assertEqual(auto_notation(123, 4), '123.0')
    self.assertEqual(auto_notation(1234, 4), '1.234e3')

class TestEngNotation(unittest.TestCase):
  def test_multi(self):
    self.assertEqual(eng_notation(1, 1), '1e0')
    self.assertEqual(eng_notation("1", 1), '1e0') # type conversion
    self.assertEqual(eng_notation(1, 2), '1.0e0')
    self.assertEqual(eng_notation(10, 1), '10e0')

    self.assertEqual(eng_notation(123, 1), '100e0')
    self.assertEqual(eng_notation(123, 2), '120e0')  # round down
    self.assertEqual(eng_notation(-1260, 2), '-1.3e3')  # round up

    self.assertEqual(eng_notation(.123, 1), '100e-3')
    self.assertEqual(eng_notation(-.0123, 2), '-12e-3')
    self.assertEqual(eng_notation(-.0123, 3), '-12.3e-3')
    self.assertEqual(eng_notation(.126, 2), '130e-3')

    self.assertEqual(eng_notation(-123, 4), '-123.0e0')  # round down

    # sig zero
    self.assertEqual(eng_notation(10, 2), '10.e0')

class TestStdNotation(unittest.TestCase):
  def test_multi(self):
    self.assertEqual(std_notation(123, 1), '100')
    self.assertEqual(std_notation("123", 1), '100') # type conversion
    self.assertEqual(std_notation(123, 2), '120')  # round down
    self.assertEqual(std_notation(126, 2), '130')  # round up
    self.assertEqual(std_notation(1260, 2), '1300')  # round up

    self.assertEqual(std_notation(.123, 1), '0.1')
    self.assertEqual(std_notation(.0123, 2), '0.012')
    self.assertEqual(std_notation(.126, 2), '0.13')
    self.assertEqual(std_notation(.126, 4), '0.1260')

    self.assertEqual(std_notation(123, 3), '123')
    self.assertEqual(std_notation(123, 4), '123.0')  # round down

    # sig zero
    self.assertEqual(std_notation(10, 2), '10.')

class TestSciNotation(unittest.TestCase):
  def test_multi(self):
    self.assertEqual(sci_notation(1, 1), '1e0')
    self.assertEqual(sci_notation("1", 1), '1e0')  # type conversion
    self.assertEqual(sci_notation(1, 2), '1.0e0')
    self.assertEqual(sci_notation(10, 1), '1e1')

    self.assertEqual(sci_notation(123, 1), '1e2')
    self.assertEqual(sci_notation(123, 2), '1.2e2')  # round down
    self.assertEqual(sci_notation(-126, 2), '-1.3e2')  # round up

    self.assertEqual(sci_notation(.123, 1), '1e-1')
    self.assertEqual(sci_notation(-.0123, 2), '-1.2e-2')
    self.assertEqual(sci_notation(.126, 2), '1.3e-1')

    self.assertEqual(sci_notation(123, 3), '1.23e2')
    self.assertEqual(sci_notation(-123, 4), '-1.230e2')  # round down

class TestPlaceDot(unittest.TestCase):
  def test_all(self):
    self.assertEqual(_place_dot('123',   0, False, 'dot'), '123'  )
    self.assertEqual(_place_dot('120',   0, False, 'dot'), '120.' )
    self.assertEqual(_place_dot('0',     0, False, 'dot'),   '0'  )

    self.assertEqual(_place_dot('123',   2, False, 'dot'), '12300')

    self.assertEqual(_place_dot('123',  -2, False, 'dot'), '1.23'   )
    self.assertEqual(_place_dot('123',  -3, False, 'dot'), '0.123'  )
    self.assertEqual(_place_dot('123',  -5, False, 'dot'), '0.00123')

    # Zero stripping
    self.assertEqual(_place_dot('123' ,  2, True,  'dot'), '12300')
    self.assertEqual(_place_dot('100' , -1, True,  'dot'),    '10'   )
    self.assertEqual(_place_dot('1200', -2, False, 'dot'),    '12.00')
    self.assertEqual(_place_dot('1200', -2, True,  'dot'),    '12'   )
    self.assertEqual(_place_dot('1200', -1, False, 'dot'),   '120.0' )
    self.assertEqual(_place_dot('1200', -1, True,  'dot'),   '120'   )

    # Convention: None
    self.assertEqual(_place_dot('120',   0, False, 'none'), '120' )
    # Convention: Overbar
    self.assertEqual(_place_dot('120',   0, False, 'overbar'), u'120\u0305' )

class TestNumberProfile(unittest.TestCase):
  def test_positive(self):
    self.assertEqual(
        _number_profile(123, 2),
        ('12', 1, False)
      )

    self.assertEqual(
        _number_profile(123, 3),
        ('123', 0, False)
      )

  def test_negative(self):
    self.assertEqual(
        _number_profile(-123, 3),
        ('123', 0, True)
      )

    self.assertEqual(
        _number_profile(-123, 1),
        ('1', 2, True)
      )

    self.assertEqual(
        _number_profile(-12.3, 2),
        ('12', 0, True)
      )

    self.assertEqual(
        _number_profile(-12.3, 3),
        ('123', -1, True)
      )

    self.assertEqual(
        _number_profile(-.123, 3),
        ('123', -3, True)
      )

    self.assertEqual(
        _number_profile(-.123, 2),
        ('12', -2, True)
      )

  def test_zeros(self):
    self.assertEqual(
        _number_profile(0, 1),
        ('0', 0, False)
      )

    self.assertEqual(
        _number_profile(0, 3),
        ('000', -2, False)
      )

  def test_round_up(self):
    self.assertEqual(
        _number_profile(-.126, 2),
        ('13', -2, True)
      )


class TestDemonstration(unittest.TestCase):
  def test_main(self):
    '''
    Verify that the demonstration.py module won't raise exceptions.
    Does not verify that the output is well formed or correct.
    ''' 
    with silence_stdout():
      demonstration.main()


@contextmanager
def silence_stdout():
  ''' Silence stdout by redirecting it to a temporary string

  Compatible with Python 2.7 & Python 3
  '''
  if sys.version_info.major == 2:
    # Python 2: Support printing of both unicode and str
    #   objects by redirecting stdout to a codec wrapper
    #   around a cStringIO object to handle the translation.
    import cStringIO
    import codecs
    string_buffer = cStringIO.StringIO()
    redirect_object = codecs.getwriter("utf8")(string_buffer)
  else:
    # Python 3: Redirect stdout to a StringIO object
    redirect_object = StringIO()

  oldstdout = sys.stdout
  sys.stdout = redirect_object
  try:
    yield
  finally:
    sys.stdout = oldstdout

if __name__ == '__main__':
  unittest.main()


try.py
__author__ = 'William Rusnack github.com/BebeSparkelSparkel linkedin.com/in/williamrusnack williamrusnack@gmail.com'

import math


def to_precision(value, precision, notation='auto'):
  '''
  converts a value to the specified notation and precision
  value - any type that can be converted to a float
  predision - integer that is greater than zero
  notation - string
    'auto' - selects standard notation when -1000 < value < 1000 else returns scientific notation
    'sci' or 'scientific' - returns scientific notation
      ref: https://www.mathsisfun.com/numbers/scientific-notation.html
    'eng' or 'engineering' - returns engineering notation
      ref: http://www.mathsisfun.com/definitions/engineering-notation.html
    'std' or 'standard' - returns standard notation
      ref: http://www.mathsisfun.com/definitions/standard-notation.html
  '''
  value = float(value)

  if notation == 'auto':
    if -1000 < value < 1000:
      converter = std_notation
    else:
      converter = sci_notation

  elif notation in ('sci', 'scientific'):
    converter = sci_notation

  elif notation in ('eng', 'engineering'):
    converter = eng_notation

  elif notation in ('std', 'standard'):
    converter = std_notation

  else:
    raise ValueError('Unknown notation: ' + str(notation))

  return converter(value, precision)


def std_notation(value, precision):
  '''
  standard notation (US version)
  ref: http://www.mathsisfun.com/definitions/standard-notation.html

  returns a string of value with the proper precision

  ex:
    std_notation(5, 2) => 5.0
    std_notation(5.36, 2) => 5.4
    std_notation(5360, 2) => 5400
    std_notation(0.05363, 3) => 0.0536

    created by William Rusnack
      github.com/BebeSparkelSparkel
      linkedin.com/in/williamrusnack/
      williamrusnack@gmail.com
  '''
  sig_digits, power, is_neg = _number_profile(value, precision)

  return ('-' if is_neg else '') + _place_dot(sig_digits, power)


def sci_notation(value, precision, filler):
  '''
  scientific notation
  ref: https://www.mathsisfun.com/numbers/scientific-notation.html

  returns a string of value with the proper precision and 10s exponent
  filler is placed between the decimal value and 10s exponent

  ex:
    sci_notation(123, 1, 'E') => 1E2
    sci_notation(123, 3, 'E') => 1.23E2
    sci_notation(.126, 2, 'E') => 1.3E-1

    created by William Rusnack
      github.com/BebeSparkelSparkel
      linkedin.com/in/williamrusnack/
      williamrusnack@gmail.com
  '''
  is_neg, sig_digits, dot_power, ten_power = _sci_notation(value, precision)

  return ('-' if is_neg else '') + _place_dot(sig_digits, dot_power) + filler + str(ten_power)


def eng_notation(value, precision, filler):
  '''
  engineering notation
  ref: http://www.mathsisfun.com/definitions/engineering-notation.html

  returns a string of value with the proper precision and 10s exponent that is divisable by 3
  filler is placed between the decimal value and 10s exponent

  created by William Rusnack
    github.com/BebeSparkelSparkel
    linkedin.com/in/williamrusnack/
    williamrusnack@gmail.com
  '''
  is_neg, sig_digits, sci_dot, sci_power = _sci_notation(value, precision)

  eng_power = 3 * math.floor(sci_power / 3)
  eng_dot = sci_dot + sci_power - eng_power

  return ('-' if is_neg else '') + _place_dot(sig_digits, eng_dot) + filler + str(eng_power)


def _sci_notation(value, precision):
  '''
  returns the properties for to construct a scientific notation number
  used in sci_notation and eng_notation

  created by William Rusnack
    github.com/BebeSparkelSparkel
    linkedin.com/in/williamrusnack/
    williamrusnack@gmail.com
  '''
  sig_digits, power, is_neg = _number_profile(value, precision)

  dot_power = -(precision - 1)
  ten_power = power + precision - 1

  return is_neg, sig_digits, dot_power, ten_power


def _place_dot(digits, power):
  '''
  places the dot in the correct spot in the digits
  if the dot is outside the range of the digits zeros will be added

  ex:
    _place_dot(123, 2) => 12300
    _place_dot(123, -2) => 1.23
    _place_dot(123, 3) => 0.123
    _place_dot(123, 5) => 0.00123

    created by William Rusnack
      github.com/BebeSparkelSparkel
      linkedin.com/in/williamrusnack/
      williamrusnack@gmail.com
  '''
  if power > 0:
    out = digits + '0' * power

  elif power < 0:
    power = abs(power)
    precision = len(digits)

    if power < precision:
      out = digits[:-power] + '.' + digits[-power:]

    else:
      out = '0.' + '0' * (power - precision) + digits

  else:
    out = digits + ('.' if digits[-1] == '0' else '')

  return out


def _number_profile(value, precision):
  '''
  returns:
    string of significant digits
    10s exponent to get the dot to the proper location in the significant digits
    bool that's true if value is less than zero else false

    created by William Rusnack
      github.com/BebeSparkelSparkel
      linkedin.com/in/williamrusnack/
      williamrusnack@gmail.com
  '''
  if value == 0:
    sig_digits = '0' * precision
    power = -(1 - precision)
    is_neg = False

  else:
    if value < 0:
      value = abs(value)
      is_neg = True
    else:
      is_neg = False

    power = -1 * math.floor(math.log10(value)) + precision - 1
    sig_digits = str(round(abs(value) * 10**power))

  return sig_digits, -power, is_neg


to_precision.py
__author__ = '''William Rusnack github.com/BebeSparkelSparkel linkedin.com/in/williamrusnack williamrusnack@gmail.com
Eric Moyer github.com/epmoyer eric@lemoncrab.com'''

import sys
from math import floor, log10


def std_notation(value, precision):
  '''
  standard notation (US version)
  ref: http://www.mathsisfun.com/definitions/standard-notation.html

  returns a string of value with the proper precision

  ex:
    std_notation(5, 2) => '5.0'
    std_notation(5.36, 2) => '5.4'
    std_notation(5360, 2) => '5400'
    std_notation(0.05363, 3) => '0.0536'

    created by William Rusnack
      github.com/BebeSparkelSparkel
      linkedin.com/in/williamrusnack/
      williamrusnack@gmail.com
  '''
  return to_precision(value, precision, notation='std')


def sci_notation(value, precision, delimiter='e'):
  '''
  scientific notation
  ref: https://www.mathsisfun.com/numbers/scientific-notation.html

  returns a string of value with the proper precision and 10s exponent
  delimiter is placed between the decimal value and 10s exponent

  ex:
    sci_notation(123, 1, 'E') => '1E2'
    sci_notation(123, 3, 'E') => '1.23E2'
    sci_notation(.126, 2, 'E') => '1.3E-1'

    created by William Rusnack
      github.com/BebeSparkelSparkel
      linkedin.com/in/williamrusnack/
      williamrusnack@gmail.com
  '''
  return to_precision(value, precision, notation='sci', delimiter=delimiter)


def eng_notation(value, precision, delimiter='e'):
  '''
  engineering notation
  ref: http://www.mathsisfun.com/definitions/engineering-notation.html

  returns a string of value with the proper precision and 10s exponent that is divisible by 3
  delimiter is placed between the decimal value and 10s exponent

  ex:
    sci_notation(123, 1, 'E') => '100E0'
    sci_notation(1230, 3, 'E') => '1.23E3'
    sci_notation(.126, 2, 'E') => '120E-3'

  created by William Rusnack
    github.com/BebeSparkelSparkel
    linkedin.com/in/williamrusnack/
    williamrusnack@gmail.com
  '''
  return to_precision(value, precision, notation='eng', delimiter=delimiter)


def auto_notation(value, precision, delimiter='e'):
  '''
  Automatically selects between standard notation (US version) and scientific notation.
  Values in the range 0.001 < abs(value) < 1000 return standard notation.

  http://www.mathsisfun.com/definitions/standard-notation.html
  https://www.mathsisfun.com/numbers/scientific-notation.html

  returns a string of value with the proper precision

  ex:
    auto_notation(123, 4) => '123.4'
    std_notation(1234, 4) => '1.234e3'
  '''
  return to_precision(value, precision, notation='auto', delimiter=delimiter)


def to_precision(
  value,
  precision,
  notation='auto',
  delimiter='e',
  auto_limit=3,
  strip_zeros=False,
  preserve_integer=False,
  convention='dot'):
  '''
  converts a value to the specified notation and precision
  value - any type that can be converted to a float
  precision - integer that is greater than zero
  notation - string
    'auto' - selects standard notation when abs(power) < auto_limit else
      returns scientific notation.
    'sci' or 'scientific' - returns scientific notation
      ref: https://www.mathsisfun.com/numbers/scientific-notation.html
    'eng' or 'engineering' - returns engineering notation
      ref: http://www.mathsisfun.com/definitions/engineering-notation.html
    'std' or 'standard' - returns standard notation
      ref: http://www.mathsisfun.com/definitions/standard-notation.html
  delimiter - is placed between the decimal value and 10s exponent
  auto_limit - integer. When abs(power) exceeds this limit, 'auto'
    mode will return scientific notation.
  strip_zeros - if true, trailing decimal zeros will be removed.
  preserve_integer - if true, 'std' will preserve all digits when returning
    values that have no decimal component.
  convention - string
    'dot' - selects dot convention
      ref: https://en.wikipedia.org/wiki/Significant_figures#Significant_figures_rules_explained
    'overbar' - selects overbar convention
      ref: https://en.wikipedia.org/wiki/Significant_figures#Significant_figures_rules_explained
  '''
  is_neg, sig_digits, dot_power, ten_power = _sci_decompose(value, precision)

  if notation == 'auto':
    if abs(ten_power) < auto_limit:
      converter = _std_notation
    else:
      converter = _sci_notation

  elif notation in ('sci', 'scientific'):
    converter = _sci_notation

  elif notation in ('eng', 'engineering'):
    converter = _eng_notation

  elif notation in ('std', 'standard'):
    converter = _std_notation

  else:
    raise ValueError('Unknown notation: ' + str(notation))

  return converter(value, precision, delimiter, strip_zeros,
                   preserve_integer, convention)


def _std_notation(value, precision, _, strip_zeros, preserve_integer, convention):
  '''
  standard notation (US version)
  ref: http://www.mathsisfun.com/definitions/standard-notation.html

  returns a string of value with the proper precision

  strip_zeros - if true, trailing decimal zeros will be removed.
  preserve_integer - if true, 'std' will preserve all digits when returning
    values that have no decimal component.
  '''
  sig_digits, power, is_neg = _number_profile(value, precision)
  result = ('-' if is_neg else '') + _place_dot(sig_digits, power, strip_zeros, convention)

  if preserve_integer and not '.' in result:
    # Result was an integer, preserve all digits
    result = '{:0.0f}'.format(value)

  return result


def _sci_notation(value, precision, delimiter, strip_zeros, _, convention):
  '''
  scientific notation
  ref: https://www.mathsisfun.com/numbers/scientific-notation.html

  returns a string of value with the proper precision and 10s exponent
  delimiter is placed between the decimal value and 10s exponent

  strip_zeros - if true, trailing decimal zeros will be removed.
  '''
  is_neg, sig_digits, dot_power, ten_power = _sci_decompose(value, precision)

  return (('-' if is_neg else '') +
          _place_dot(sig_digits, dot_power, strip_zeros, convention) +
          delimiter +
          str(ten_power))

def _eng_notation(value, precision, delimiter, strip_zeros, _, convention):
  '''
  engineering notation
  ref: http://www.mathsisfun.com/definitions/engineering-notation.html

  returns a string of value with the proper precision and 10s exponent that is divisible by 3
  delimiter is placed between the decimal value and 10s exponent

  strip_zeros - if true, trailing decimal zeros will be removed.
  '''
  is_neg, sig_digits, dot_power, ten_power = _sci_decompose(value, precision)

  eng_power = int(3 * floor(ten_power / 3))
  eng_dot = dot_power + ten_power - eng_power

  return (('-' if is_neg else '') + 
          _place_dot(sig_digits, eng_dot, strip_zeros, convention) +
          delimiter +
          str(eng_power))


def _sci_decompose(value, precision):
  '''
  returns the properties for to construct a scientific notation number
  used in sci_notation and eng_notation

  created by William Rusnack
    github.com/BebeSparkelSparkel
    linkedin.com/in/williamrusnack/
    williamrusnack@gmail.com
  '''
  value = float(value)
  sig_digits, power, is_neg = _number_profile(value, precision)

  dot_power = -(precision - 1)
  ten_power = power + precision - 1

  return is_neg, sig_digits, dot_power, ten_power


def _place_dot(digits, power, strip_zeros, convention):
  '''
  places the dot in the correct spot in the digits
  if the dot is outside the range of the digits zeros will be added
  if strip_zeros is set, trailing decimal zeros will be removed

  ex:
    _place_dot('123',   2, False) => '12300'
    _place_dot('123',  -2, False) => '1.23'
    _place_dot('123',   3, False) => '0.123'
    _place_dot('123',   5, False) => '0.00123'
    _place_dot('120',   0, False) => '120.'
    _place_dot('1200', -2, False) => '12.00'
    _place_dot('1200', -2, True ) => '12'
    _place_dot('1200', -1, False) => '120.0'
    _place_dot('1200', -1, True ) => '120'

    created by William Rusnack
      github.com/BebeSparkelSparkel
      linkedin.com/in/williamrusnack/
      williamrusnack@gmail.com
  '''
  #print(f'_place_dot(digits={digits}, power={power}, strip_zeros={strip_zeros}')

  if convention not in ('overbar', 'dot', 'none'):
    raise ValueError('Unknown convention: ' + str(convention))

  if power > 0:
    if convention == 'overbar':
      # Add overbar to last digit before added (non-significant) zeros
      out = digits + u'\u0305' + '0' * power
    else:
      # Convention is dot or none
      out = digits + '0' * power

  elif power < 0:
    power = abs(power)
    precision = len(digits)

    if power < precision:
      out = digits[:-power] + '.' + digits[-power:]

    else:
      out = '0.' + '0' * (power - precision) + digits

  else:
    if convention == 'overbar':
      # Add overbar to last digit if representation is non-zero and ends in zero
      out = digits + (u'\u0305' if digits[-1] == '0' and len(digits) > 1 else '')
    elif convention == 'dot':
      # Add trailing decimal if representation is non-zero and ends in zero
      out = digits + ('.' if digits[-1] == '0' and len(digits) > 1 else '')
    else:
      out = digits

  if strip_zeros and '.' in out:
    out = out.rstrip('0').rstrip('.')

  if convention == 'overbar' and sys.version_info.major == 2:
    # Always return unicode on Python 2 in overbar mode (rather than
    # only when an overbar appears)
    out = unicode(out)

  return out


def _number_profile(value, precision):
  '''
  returns:
    string of significant digits
    10s exponent to get the dot to the proper location in the significant digits
    bool that's true if value is less than zero else false

    created by William Rusnack
      github.com/BebeSparkelSparkel
      linkedin.com/in/williamrusnack/
      williamrusnack@gmail.com
  '''
  value = float(value)
  if value == 0:
    sig_digits = '0' * precision
    power = -(1 - precision)
    is_neg = False

  else:
    if value < 0:
      value = abs(value)
      is_neg = True
    else:
      is_neg = False

    power = -1 * floor(log10(value)) + precision - 1
    sig_digits = str(int(round(abs(value) * 10.0**power)))

  return sig_digits, int(-power), is_neg

